patches/server/1068-Memory-optimization-Lazy-initialization-of-IBlockDat.patch

```diff
@@ -0,0 +1,95 @@
+@@ -4,6 +4,8 @@ import ca.spottedleaf.moonrise.common.list.IBlockDataList;
+ 
+ public interface BlockCountingChunkSection {
+ 
++    public boolean anyTickingBlocks();
++
+     public int moonrise$getSpecialCollidingBlocks();
+ 
+     public IBlockDataList moonrise$getTickingBlockList();
+diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
+index f9abf63e12ea930275121b470e4e4906cff0fc12..010c2969bf5fa5bb7281acb3cdb44f3193e086e7 100644
+--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
++++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
+@@ -817,6 +817,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
+                 continue;
+             }
+ 
++            if (!((ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection)section).anyTickingBlocks()) {
++                continue;
++            }
++
+             final ca.spottedleaf.moonrise.common.list.IBlockDataList tickList = ((ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection)section).moonrise$getTickingBlockList();
+             if (tickList.size() == 0) {
+                 continue;
+diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
+index c3b1caa352b988ec44fa2b2eb0536517711f5460..3ae68c1f7eb039b7a7abb84ade06737bdf03453f 100644
+--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
++++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
+@@ -35,7 +35,16 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
+     }
+ 
+     private int specialCollidingBlocks;
+-    private final ca.spottedleaf.moonrise.common.list.IBlockDataList tickingBlocks = new ca.spottedleaf.moonrise.common.list.IBlockDataList();
++    private ca.spottedleaf.moonrise.common.list.IBlockDataList tickingBlocks = null;
++
++    @Override
++    public boolean anyTickingBlocks() {
++        return tickingBlocks != null;
++    }
++
++    private void createTickingBlocks() {
++        if (tickingBlocks == null) tickingBlocks = new ca.spottedleaf.moonrise.common.list.IBlockDataList();
++    }
+ 
+     @Override
+     public final int moonrise$getSpecialCollidingBlocks() {
+@@ -44,6 +53,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
+ 
+     @Override
+     public final ca.spottedleaf.moonrise.common.list.IBlockDataList moonrise$getTickingBlockList() {
++        this.createTickingBlocks();
+         return this.tickingBlocks;
+     }
+     // Paper end - block counting
+@@ -126,9 +136,11 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
+         }
+ 
+         if (iblockdata1.isRandomlyTicking()) {
++            this.createTickingBlocks();
```

Comment:
why create list when just trying to remove an entry?

Reply:
Because if `this.tickingBlocks` is null, a call to `this.tickingBlocks.remove()` would crash immediately.

In theory there shouldn't really be any calls to `remove` when there are no tickingBlocks... But I decided to create the structure, just to not modify the original behavior.

It could be changed to `if (this.tickingBlocks != null) this.tickingBlocks.add(x, y, z, state);`, but I decided just creating it is simpler (1 loc vs 3 loc).